rclcpp/src/rclcpp/callback_group.cpp

```diff
@@ -23,6 +23,13 @@ CallbackGroup::CallbackGroup(CallbackGroupType group_type)
 : type_(group_type), can_be_taken_from_(true)
 {}
 
+const std::vector<rclcpp::PublisherBase::WeakPtr> &
+CallbackGroup::get_publisher_ptrs() const
+{
+  std::lock_guard<std::mutex> lock(mutex_);
```

Comment:
I believe this logic is incorrect, we are locking a mutex_, then returning a const reference. The mutex unlocks at the end of this context (since it is a lock guard). The caller can read from the exact vector, which means this mutex is not protecting anything.

Reply:
Good point. But this pattern is also found in the other original functions, like `get_subscription_ptrs()`, `get_timer_ptrs()`, etc. I think it is to prevent `get_publisher_ptrs()` from returning something corrupted when `add_publisher()` is also being called.

Reply:
If the pattern is broken we can introduce a new pattern. What is the actual need to expose the list of pointers? Could that be wrapped in a function inside CallbackGroup where it is protected by the mutex?

Reply:
What would be the best way to fix? std::vector isn't thread safe and returning a copy could be expensive.

Reply:
The fix depends on what it's actually used for. 

One fix would be to look at implementing an iterator that would become invalid if the data structure was changed while iterating

Reply:
I don't disagree with the points being made here, but I'd like to avoid the introducing the temptation to go fix the remainder of the implementation. I'd suggest we log an issue against rclcpp that this pattern isn't safe - use the pattern - and fix it in a separate targeted change

Reply:
I think if the `Waitable`s refactoring is outside the scope of this PR, then so should be changing this locking pattern, because at least the `Waitable`s refactoring wasn't supposed to introduce functional changes, but changing this would be a functional change.

I also think the presence of this mutex could be a red herring; it being there makes it look like the vector will be accessed by multiple threads, when it might not actually ever be and this broken pattern has been working the entire time.

Reply:
I think we're all in agreement that we're not going to introduce a different pattern for now, but would follow up with an issue or other pull request. 

Resolving the conversation.